spawnpoint.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 Siôn le Roux.  All rights reserved.
+// Use of this source code is subject to an MIT-style
+// licence which can be found in the LICENSE file.
+
+package main
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/solarlune/resolv"
+)
+
+// spawnDistance is the distance where the point is activated
+const spawnDistance = 250
+
+// SpawnPoints is an array of SpawnPoint
+type SpawnPoints []*SpawnPoint
+
+// Update updates all the SpawnPoints
+func (sps *SpawnPoints) Update(g *Game) {
+	for _, s := range *sps {
+		s.Update(g)
+	}
+}
+
+// SpawnPosition describes the spawning position related to the SpawnPoint center
+type SpawnPosition struct {
+	Distance int // Distance of the position from the center
+	Angle    int // Angle of the position
+}
+
+// SpawnPoint is a point on the map where zombies are spawn
+type SpawnPoint struct {
+	Position       Coord
+	InitialCount   int
+	Continuous     bool
+	Zombies        Zombies
+	InitialSpawned bool
+	PrevPosition   SpawnPosition
+	NextSpawn      int
+}
+
+// NextPosition gives the offset of the next spawning to the center of the point
+func (s *SpawnPoint) NextPosition() Coord {
+	
+	// Move further if zombies have been spwaned around the whole circle
+	if s.PrevPosition.Angle == 0 {
+		s.PrevPosition.Distance++
+	}
+
+	// Spawn positions in every 60 degress (360 / 6)
+	s.PrevPosition.Angle = (s.PrevPosition.Angle + 1) % 6
+
+	return Coord{
+		X: math.Cos(-2*math.Pi / 6 * float64(s.PrevPosition.Angle)) * float64(s.PrevPosition.Distance),
+		Y: math.Sin(-2*math.Pi / 6 * float64(s.PrevPosition.Angle)) * float64(s.PrevPosition.Distance),
+	}
+}
+
+// SpawnZombie spawns one zombie
+func (s *SpawnPoint) SpawnZombie(g *Game) {
+	np := s.NextPosition()
+
+	zombieType := rand.Intn(zombieTypes)
+	speed := zombieSpeed
+	if zombieType == 2 {
+		speed = zombieRunnerSpeed
+	} else if zombieType == 3 {
+		speed = zombieCrawlerSpeed
+	}
+
+	sprites := g.ZombieSprites[zombieType]
+	dimensions := sprites.Sprite[0].Position
+	z := &Zombie{
+		Object: resolv.NewObject(
+			s.Position.X + np.X * 32, s.Position.Y + np.Y * 32,
+			float64(dimensions.W), float64(dimensions.H),
+			tagMob,
+		),
+		Angle:    0,
+		Sprite:   sprites,
+		Speed:    speed * (1 + rand.Float64()),
+		HitToDie: 1 + rand.Intn(2),
+	}
+	z.Object.Data = z // self-reference for later
+	z.Target = g.Player.Object
+	z.SpawnPoint = s
+
+	g.Space.Add(z.Object)
+	g.Zombies = append(g.Zombies, z)
+	s.Zombies = append(s.Zombies, z)
+	s.NextSpawn = 180 + rand.Intn(180)
```

Comment:
The trick is: NextSpawn is used only in case of continuous spawning, not in initial. Thus the first batch of zombies are spawned immediately together and then the new ones will come in 3-6 secs after each other

Reply:
Oh I see... that's quite clever 👌